File: service/python/v2/main.py
import time
import requests
from functools import partial
from flask import Flask, jsonify, g, request
from multiprocessing.dummy import Pool as ThreadPool


app = Flask(__name__)


def getForwardHeaders(request):
    headers = {}
    incoming_headers = [
        'x-request-id',
        'x-b3-traceid',
        'x-b3-spanid',
        'x-b3-parentspanid',
        'x-b3-sampled',
        'x-b3-flags',
        'x-ot-span-context'
    ]

    for ihdr in incoming_headers:
        val = request.headers.get(ihdr)
        if val is not None:
            headers[ihdr] = val
            # print("incoming: "+ihdr+":"+val)

    return headers


@app.before_request
def before_request():
    g.forwardHeaders = getForwardHeaders(request)


def get_url_response(url, headers={}):
    try:
        start = time.time()
        resp = requests.get(url, headers=headers, timeout=20)
        response_time = round(time.time() - start, 2)
        data = resp.json()
        data['response_time'] = response_time
    except Exception as e:
        print(e)
        data = None
    return data


@app.route("/env")
def env():
    # service_lua_url = 'http://httpbin.org/delay/3'
    # service_node_url = 'http://httpbin.org/delay/4'

    service_lua_url = 'http://' + 'service-lua' + '/env'
    service_node_url = 'http://' + 'service-node' + '/env'

    services_url = [service_lua_url, service_node_url]
    pool = ThreadPool(2)
    wrap_get_url_response = partial(get_url_response, headers=g.forwardHeaders)
    results = pool.map(wrap_get_url_response, services_url)
    upstream = [r for r in results if r]

    return jsonify({
        "message": 'python v1',
        "upstream": upstream
    })


@app.route("/status")
def status():
    return "ok"


if __name__ == '__main__':
    app.run(host='0.0.0.0', port=80)

File: service/node/v2/main.js
const Koa = require('koa');
const Router = require('koa-router');
const axios = require('axios')
const app = new Koa();
const router = new Router();


function getForwardHeaders(request) {
    headers = {}
    incoming_headers = [
        'x-request-id',
        'x-b3-traceid',
        'x-b3-spanid',
        'x-b3-parentspanid',
        'x-b3-sampled',
        'x-b3-flags',
        'x-ot-span-context'
    ]

    for (idx in incoming_headers) {
        ihdr = incoming_headers[idx]
        val = request.headers[ihdr]
        if (val !== undefined && val !== '') {
            headers[ihdr] = val
            console.log("incoming: " + ihdr + ":" + val)
        }
    }
    return headers
}



router.get('/status', async (ctx, next) => {
    ctx.body = 'ok';
})

router.get('/env', async (ctx, next) => {
    forwardHeaders = getForwardHeaders(ctx.request)
    // service_go_url = 'http://httpbin.org/delay/10'
    service_go_url = 'http://' + 'service-go' + '/env'
    upstream_ret = null
    try {
        // console.log(forwardHeaders)
        // upstream_ret = JSON.parse('{"message": "go v1"}')
        let start = Date.now()
        const response = await axios.get(service_go_url, {
            headers: forwardHeaders,
            timeout: 20000
        });
        response_time = ((Date.now() - start) / 1000).toFixed(2)
        upstream_ret = response.data
        upstream_ret.response_time = response_time
    } catch (error) {
        console.error('error');
    }
    if (upstream_ret) {
        // console.log(upstream_ret);
        ctx.body = {
            'message': 'node v2',
            'upstream': [upstream_ret]
        };
    } else {
        ctx.body = {
            'message': 'node v2',
            'upstream': []
        }
    }
})

app.use(router.routes()).use(router.allowedMethods());
app.listen(80);